#580 Add coveralls support

rmjarvis opened the pull request:
I've added coverage testing and reporting via coveralls to the travis config file.  I also added a program that does the equivalent of `check_yaml` for Travis to run, so it can check those as well, and also count the config lines that are run there for the coverage.

I don't know whether this should close #580 or not.  The coveralls report seems most likely to induce people to expand the coverage of the unit tests than an old issue sitting on the second page of the issues list.  So I'm inclined to close it after this.  But if people think it's useful to have an issue that explicitly requests more unit test coverage, we can leave it open.

rmandelb commented:
Possibly I am doing something wrong (?), but when I go to this branch and try to do the usual ```scons``` / ```scons install``` / ```scons tests```, I get a failure of ```scons tests```:
```
Starting python tests...
bash: /Library/Frameworks/EPD64.framework/Versions/Current/bin/nosetests test*.py: No such file or directory
Nosetests returned error code  127
scons: *** [tests/tests.log] Explicit exit, status 127
scons: building terminated because of errors.
```
Perhaps it's not going into the tests/ directory before running this command, and that's why it's not finding any ```test*.py``` files?

rmandelb commented:
That is very cool.  

> Should we add the coveralls badge (and the travis CI badge, and an arxiv badge) to the top 
> of README.md?

+1

> I don't know whether this should close #580 or not.

Given that it appears way down the list, so it mostly isn't seen, I'd go with closing it.  The coveralls report will indeed be a more useful reminder.

jmeyers314 commented on tests/run_examples.py:
```diff
@@ -0,0 +1,304 @@
+
+
+if __name__ == "__main__":
+    shutil.rmtree('../examples/output')
```
Should probably make this path with respect to `__file__`, in case someone starts this script from a directory other than $GALSIM_DIR/tests/.

Also, I think this line crashes if the target directory doesn't exist.

rmjarvis replied:
Good points.  Fixed.

jmeyers314 commented:
One of these days I should learn how the config framework works, so I can give that a proper review.... (I did scan it briefly).  But until then, this looks good to me, just the one very minor point above.

rmjarvis commented:
OK, I think I'm done futzing with this now.  I added a .coveragerc file to have it ignore various kinds of things that we don't actually care about in terms of unit test coverage.  And for some specific bits of code (special cases for old systems, obsolete code that's still sitting there, warnings about bad user input, etc.) I marked them with a `pragma: no cover` comment to exclude them from the coverage report.

I also added code to run the des and great3 examples, which covers a bit more of the code base, at least for syntax errors, if not actual correctness testing, although I think for the additional bits that these catch, syntax errors are probably the main thing we'd worry about.

After all that, the remaining bits I think are code that ought to be included in the unit tests, and of them, we're currently at 93% coverage.  Not too shabby, IMHO.  You can peruse them [here](https://coveralls.io/github/GalSim-developers/GalSim).  It's pretty cool -- the covered lines are green, and the uncovered lines are pink.

I even felt inspired to add a unit test for the `NFWHalo.getLensing` function that was completely untested.  And in doing so, I discovered a bug in it, so that was useful.  I also found a number of typos in uncovered bits of code (like writing `NotImplemented` rather than `NotImplementedError`), so I fixed them as well.

jmeyers314 commented:
Very nice.

Is it straightforward to cherry-pick some of these onto 1.4 so we can have the badges show up on the default github landing page?

rmjarvis commented:
> Is it straightforward to cherry-pick some of these onto 1.4

Done.

rmjarvis commented:
OK to merge this?  Or did anyone want to look through my choices for things to skip coverage on?  

I know @jmeyers314 you looked through the code early on, but then I added a bunch of `pragma: no cover` comments, which I don't know if you care about reviewing.

rmandelb commented:
I'm fine with merging.

jmeyers314 commented:
Me too.
